sync15-adapter/src/lib.rs

```diff
@@ -0,0 +1,341 @@
+            error!("HTTP error {} ({}) during storage request to {}",
+                   resp.status().as_u16(), resp.status(), resp.url().path());
+            bail!(error::ErrorKind::StorageHttpError(
+                resp.status(), resp.url().path().into()));
+        }
+
+        // TODO:
+        // - handle backoff
+        // - x-weave-quota?
+        // - ... almost certainly other things too...
+
+        Ok(resp)
+    }
+
+    fn collection_request(&self, method: Method, r: &CollectionRequest) -> error::Result<Response> {
+        self.make_storage_request(method.clone(),
+                                  r.build_url(Url::parse(&self.tsc.token().api_endpoint)?)?)
+    }
+
+    fn fetch_info<T>(&self, path: &str) -> error::Result<T> where for <'a> T: serde::de::Deserialize<'a> {
+        let mut resp = self.relative_storage_request(Method::Get, path)?;
+        let result: T = resp.json()?;
+        Ok(result)
+    }
+
+    pub fn remote_setup(&mut self) -> error::Result<()> {
+        let server_config = self.fetch_info::<InfoConfiguration>("info/configuration")?;
+        self.server_config = Some(server_config);
+        let mut resp = match self.relative_storage_request(Method::Get, "storage/meta/global") {
+            Ok(r) => r,
+            // This is gross, but at least it works. Replace 404s on meta/global with NoMetaGlobal.
+            Err(error::Error(error::ErrorKind::StorageHttpError(hyper::StatusCode::NotFound, ..), _)) =>
+                bail!(error::ErrorKind::NoMetaGlobal),
+            Err(e) => return Err(e),
+        };
+        // Note: meta/global is not encrypted!
+        let meta_global: BsoRecord<MetaGlobalRecord> = resp.json()?;
+        info!("Meta global: {:?}", meta_global.payload);
+        let collections = self.fetch_info::<HashMap<String, ServerTimestamp>>("info/collections")?;
+        self.update_keys(&collections)?;
+        self.last_sync_remote = collections;
+        Ok(())
+    }
+
+    fn update_keys(&mut self, _info_collections: &HashMap<String, ServerTimestamp>) -> error::Result<()> {
+        // TODO: if info/collections says we should, upload keys.
+        // TODO: This should be handled in collection_keys.rs, which should track modified time, etc.
+        let mut keys_resp = self.relative_storage_request(Method::Get, "storage/crypto/keys")?;
+        let keys: BsoRecord<EncryptedPayload> = keys_resp.json()?;
+        self.keys = Some(CollectionKeys::from_encrypted_bso(keys, &self.root_key)?);
+        // TODO: error handling... key upload?
+        Ok(())
+    }
+
+    pub fn key_for_collection(&self, collection: &str) -> error::Result<&KeyBundle> {
+        Ok(self.keys.as_ref()
+                    .ok_or_else(|| error::unexpected("Don't have keys (yet?)"))?
+                    .key_for_collection(collection))
+    }
+
+    pub fn all_records<T>(&mut self, collection: &str) ->
+            error::Result<Vec<BsoRecord<MaybeTombstone<T>>>> where T: Sync15Record {
```

Comment:
Consider whether it's better to do this with generics or with trait objects. You're using monomorphization here, and that means there are separate `all_records` implementations for every `T`. Presumably the set of methods will grow, and the set of record types will grow to 5+…

Another decision you can make is to have the bit that depends on the record type — decryption to cleartext — be inside the generic method, and the rest (the first three lines) be in a separate non-generic method. That'll keep code size smaller.

Reply:
I don't expect this API to live that long FWIW, and if you exclude deserializing after decryption, this code is not very big, even with the deserialization it's only 7k or so after assembling for the password records.